realm/realm-library/src/objectServer/java/io/realm/PermissionManager.java

```diff
@@ -0,0 +1,388 @@
+/*
+ * Copyright 2017 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm;
+
+import android.os.Handler;
+
+import java.io.Closeable;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+
+import io.realm.log.RealmLog;
+import io.realm.internal.permissions.ManagementModule;
+import io.realm.internal.permissions.PermissionModule;
+
+
+/**
+ * FIXME: Better Javadoc
+ * Helper class for interacting with Realm permissions.
+ * This class has connections to underlying Realms, so all data coming from this class is thread-confined and must be
+ * closed after use to avoid leaking resources.
+ */
+public class PermissionManager implements Closeable {
+
+    private enum RealmType {
+        PERMISSION_REALM("__permission"),
+        MANAGEMENT_REALM("__management");
+
+        private final String name;
+
+        RealmType(String realmName) {
+            this.name = realmName;
+        }
+
+        public String getName() {
+            return name;
+        }
+    }
+
+    // Used to track the lifecycle of the PermissionManager
+    private RealmAsyncTask managementRealmOpenTask;
+    private RealmAsyncTask permissionRealmOpenTask;
+    private boolean openInProgress = false;
+    private boolean closed;
+
+    private final long threadId;
+    private Handler handler = new Handler();
+    private final SyncConfiguration managementRealmConfig;
+    private final SyncConfiguration permissionRealmConfig;
+    private Realm permissionRealm;
+    private Realm managementRealm;
+
+    // Task list used to queue tasks until the underlying Realms are done opening (or fail doing so).
+    private Deque<PermissionManagerAsyncTask> delayedTasks = new LinkedList<>();
+
+    // List of tasks that are running. Used to keep strong references for listeners to work.
```

Comment:
They are not running in the operating system meaning of the word, right? I don't have any better suggestions.